pkg/validation/attributes/attributes_component.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	apiAttributes "github.com/devfile/api/v2/pkg/attributes"
+)
+
+// ValidateAndReplaceForComponents validates the components data for global attribute references and replaces them with the attribute value
+func ValidateAndReplaceForComponents(attributes apiAttributes.Attributes, components []v1alpha2.Component) error {
+
+	for i := range components {
+		var err error
+
+		// Validate various component types
+		switch {
+		case components[i].Container != nil:
+			if err = validateAndReplaceForContainerComponent(attributes, components[i].Container); err != nil {
+				return err
+			}
+		case components[i].Kubernetes != nil:
+			if err = validateAndReplaceForKubernetesComponent(attributes, components[i].Kubernetes); err != nil {
+				return err
+			}
+		case components[i].Openshift != nil:
+			if err = validateAndReplaceForOpenShiftComponent(attributes, components[i].Openshift); err != nil {
+				return err
+			}
+		case components[i].Volume != nil:
+			if err = validateAndReplaceForVolumeComponent(attributes, components[i].Volume); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateAndReplaceForContainerComponent validates the container component data for global attribute references and replaces them with the attribute value
+func validateAndReplaceForContainerComponent(attributes apiAttributes.Attributes, container *v1alpha2.ContainerComponent) error {
+	var err error
+
+	if container != nil {
+		// Validate container image
+		if container.Image, err = validateAndReplaceDataWithAttribute(container.Image, attributes); err != nil {
+			return err
+		}
+
+		// Validate container commands
+		for i := range container.Command {
+			if container.Command[i], err = validateAndReplaceDataWithAttribute(container.Command[i], attributes); err != nil {
+				return err
+			}
+		}
+
+		// Validate container args
+		for i := range container.Args {
+			if container.Args[i], err = validateAndReplaceDataWithAttribute(container.Args[i], attributes); err != nil {
+				return err
+			}
+		}
+
+		// Validate memory limit
+		if container.MemoryLimit, err = validateAndReplaceDataWithAttribute(container.MemoryLimit, attributes); err != nil {
+			return err
+		}
+
+		// Validate memory limit
+		if container.MemoryRequest, err = validateAndReplaceDataWithAttribute(container.MemoryRequest, attributes); err != nil {
+			return err
+		}
+
+		// Validate source mapping
+		if container.SourceMapping, err = validateAndReplaceDataWithAttribute(container.SourceMapping, attributes); err != nil {
+			return err
+		}
+
+		// Validate container env
+		if len(container.Env) > 0 {
+			if err = validateAndReplaceForEnv(attributes, container.Env); err != nil {
+				return err
+			}
+		}
+
+		// Validate container volume mounts
+		for i := range container.VolumeMounts {
+			if container.VolumeMounts[i].Name, err = validateAndReplaceDataWithAttribute(container.VolumeMounts[i].Name, attributes); err != nil {
```

Comment:
This has the same problem as command components; `volumeMount.Name` has to refer to a volume component in the DevWorkspace, so any attribute here would have a fixed value (or else the DevWorkspace is invalid).

Reply:
makes sense, this was an oversight after your comment regd events in the prev feedback